src/ImmersedBoundaries/immersed_boundary_condition.jl

```diff
@@ -0,0 +1,290 @@
+end
+
+"""
+    ImmersedBoundaryCondition(; interfaces...)
+
+Return an ImmersedBoundaryCondition with conditions on individual
+cell `interfaces ∈ (west, east, south, north, bottom, top)`
+between the fluid and immersed boundary.
+"""
+function ImmersedBoundaryCondition(; west = nothing,
+                                     east = nothing,
+                                     south = nothing,
+                                     north = nothing,
+                                     bottom = nothing,
+                                     top = nothing)
+
+    return ImmersedBoundaryCondition(west, east, south, north, bottom, top)
+end
+
+#####
+##### Nothing and FluxBoundaryCondition.
+#####
+##### Very Important Note: For FluxBoundaryCondition,
+##### we assume fluxes are directed along the "inward-facing normal".
+##### For example, east_immersed_flux = - user_flux.
+##### With this convention, positive fluxes _increase_ boundary-adjacent
+##### cell values, and negative fluxes _decrease_ them.
+#####
+
+for side in (:west, :south, :bottom)
+    get_flux = Symbol(side, :_ib_flux)
+    @eval begin
+        @inline $get_flux(i, j, k, ibg, ::Nothing, args...) = zero(eltype(ibg))
+        @inline $get_flux(i, j, k, ibg, bc::FBC, loc, c, closure, K, id, args...) = + getbc(bc, i, j, k, ibg, args...)
+    end
+end
+
+for side in (:east, :north, :top)
+    get_flux = Symbol(side, :_ib_flux)
+    @eval begin
+        @inline $get_flux(i, j, k, ibg, ::Nothing, args...) = zero(eltype(ibg))
+        @inline $get_flux(i, j, k, ibg, bc::FBC, loc, c, closure, K, id, args...) = - getbc(bc, i, j, k, ibg, args...)
+    end
+end
+
+#####
+##### Value boundary condition. It's harder!
+#####
+
+# Harder in some ways... ValueBoundaryCondition...
+const VBC = BoundaryCondition{Value}
+const GBC = BoundaryCondition{Gradient}
+const ASD = AbstractScalarDiffusivity
+
+# "Gradient" utility for Value or Gradient boundary conditions
+@inline right_gradient(i, j, k, ibg, κ, Δ, bc::GBC, c, clock, fields) = getbc(bc, i, j, k, ibg, clock, fields)
+@inline left_gradient(i, j, k, ibg, κ, Δ, bc::GBC, c, clock, fields)  = getbc(bc, i, j, k, ibg, clock, fields)
+
+@inline function right_gradient(i, j, k, ibg, κ, Δ, bc::VBC, c, clock, fields)
+    cᵇ = getbc(bc, i, j, k, ibg, clock, fields)
+    cⁱʲᵏ = @inbounds c[i, j, k]
+    return 2 * (cᵇ - cⁱʲᵏ) / Δ
+end
+
+@inline function left_gradient(i, j, k, ibg, κ, Δ, bc::VBC, c, clock, fields)
+    cᵇ = getbc(bc, i, j, k, ibg, clock, fields)
+    cⁱʲᵏ = @inbounds c[i, j, k]
+    return 2 * (cⁱʲᵏ - cᵇ) / Δ
+end
+
+# Metric and index gymnastics for the 6 facets of the cube
+
+@inline function west_ib_flux(i, j, k, ibg, bc::VBC, (LX, LY, LZ), c, closure::ASD, K, id, clock, fields)
+    Δ = Δx(index_left(i, LX), j, k, ibg, LX, LY, LZ)
+    κ = h_diffusivity(i, j, k, ibg, flip(LX), LY, LZ, closure, K, id, clock)
+    ∇c = left_gradient(i, j, k, ibg, κ, Δ, bc, c, clock, fields)
+    return - κ * ∇c
```

Comment:
@simone-silvestri take a look at this code and also `right_gradient` and `left_gradient` to see how we satisfy `ValueBoundaryCondition` for immersed boundaries